src/base/emu-i386/simx86/interp.c

```diff
@@ -399,15 +399,15 @@ unsigned int Interp86(unsigned int PC, int mod0)
 
 static unsigned int _Interp86(unsigned int PC, int basemode)
 {
-	volatile unsigned int P0 = PC; /* volatile because of sigsetjmp */
+	volatile unsigned int P0 = PC; /* volatile because of setjmp */
 	unsigned char opc;
 	unsigned short ocs = TheCPU.cs;
 	unsigned int temp;
 	register int mode;
 	int NewNode;
 
-	if (CONFIG_CPUSIM && PROTMODE() && sigsetjmp(jmp_env, 1)) {
-		/* long jump to here from page fault */
+	if (CONFIG_CPUSIM && PROTMODE() && setjmp(jmp_env)) {
```

Comment:
I understand this is the most straight-forward
conversion, but how difficult would it be to
move the check to top level, and hence avoid
longjmp? Should then be a bit faster? I guess
with some macro tricks in interp this can be
done, what do you think?

Reply:
I was half-expecting this comment :)

It can be done but it looks fairly hard and cumbersome and certainly won't gain any speed, as the `longjmp` is exceptional (only DPMI apps that catch their own page faults really).

A read_byte would need to return some fake value (e.g. 0) then all read-modify-write instructions need to trap the read and not the write, etc. etc. so you get extra overhead on all memory instructions for an exceptional situation. Although I'm not a big fan of setjmp either I feel it's appropriate here, and note that the setjmp is outside the `while (Running)` loop.

Reply:
> A read_byte would need to return some fake value (e.g. 0)

No, I was thinking about some macro,
that will check and return, w/o calling
read_byte() (in the interp loop).
Longjmp is indeed exceptional, but you
always do the check nevertheless. The
only difference is that you will do the check
earlier, and call `return` instead of `longjmp`,
which is slightly faster, but doesn't matter
much, as it happens rare.

> A read_byte would need to return some fake value (e.g. 0) then all read-modify-write
> instructions need to trap the read and not the write, etc. etc.

Its hard to believe that you solve all these
problems with longjmp and not with plain
`return`, as IMHO its the same thing. Of
course I can believe it would require too
many refactoring, or something like that,
but this wasn't stated yes. :)

Reply:
yes the refactoring seems quite intrusive, yes you can return but then you need to do the check earlier, ie. not inside the read/write functions but before them, which means the check is no longer combined with the one for VGA/LDT... unless you remember the result and then pass that as an extra parameter to read/write... it all seems more overhead for little gain (I'm not against it per se, it just looks like a daunting task)

I was curious to see what Bochs does and they call longjmp for exceptions too:
https://github.com/svn2github/bochs/blob/master/bochs/cpu/exception.cc#L934
even if it's written in C++ and the language actually supports exceptions...

Reply:
I tried longjmp in fdpp, and that caused
leaks, because C++ dtors are not being called.
I wonder what bochs does with that...

Reply:
My guess is because Bochs uses some very restrictive dialect of C++: http://bochs.sourceforge.net/doc/docbook/development/coding.html#AEN1527

Reply:
Hmm, I guess they forgot to mention "don't
use custom destructors" or something like
that, to make longjmp()s safe.
Also with "Don't make use of any external C++ classes"
I wonder what's the use. My initial idea with
fdpp was to use header-only templates from
libstdc++, but it appears they all (or most, but
you never know beforehand) need libstdc++
to be linked in. Also I stuck with unkillable
operator delete, that comes out of nowhere,
even if you provide a custom allocator:
https://stackoverflow.com/questions/48915888/allocate-shared-with-malloc
And to avoid leaks, I had to replace longjmp
with exceptions, as there seems to be no
any unwinding library to implement c++-safe
longjumps with.
And at the end I realized that you can't do
most useful things without RTTI, and if you
try too hard, you run into bugs:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=93984
So essentially, my uncountable and persistent
attempts to find some "embedded" C++ subset,
have all miserably failed. It just seems like one
can't code up something useful on C++, unless
you enable all of its fat and heavy features.